### ISSUE TITLE
Designing an API for Imaging based Assays

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#1152]
There's been a lot of development in the imaging for drug discovery space. In particular, imaging assays are looking to become mainstream. This recent [paper](http://www.cell.com/cell-chemical-biology/fulltext/S2451-9456(18)30037-0) has been making a lot of waves. For example, check out Derek Lowe's [blog post](http://blogs.sciencemag.org/pipeline/archives/2018/03/12/images-of-machine-learning) discussion.

I propose we design an API for constructing image-based assays. We've talked about better image support for a while, but I think the time is right to execute.

@lilleswing Would love your thoughts on a potential API. What would be most useful for the pharma industry?

**Comment by joergkurtwegner** [comment]
Just in-case you need test cases, check https://idr.openmicroscopy.org

**Comment by peastman** [comment]
What do you have in mind?  The models in that paper look pretty generic, not specific to anything about images.  Are you talking about adding routines for loading image files?  Or models specifically designed for use with images?  Or pretrained models for particular applications?

**Comment by lilleswing** [comment]
The one use case I have found here is combining multiple tensors as a sample.

So each of the images has the image (obviously) but they also seem to have important metadata that is a feature.  One is a 2d matrix, ond is a 1d vector and they are of different sizes.

There are two ways of handling this right now, both are documented poorly.
1) Creating a class which holds both of them and then pickling it in a dataset and then overwriting default_generator
2) Creating a databag from multiply datasets and then overwriting default_generator

**Comment by peastman** [comment]
That would be a good thing to see if we can improve.  There's a bit of inconsistency between our models and datasets.  TensorGraph lets a model have an arbitrary number of inputs, outputs, labels, and weights.  But Dataset pretty much assumes a model only has one of each.  Can we find a clean way to let a Dataset provide multiple inputs, labels, and weights, then be able to pass that Dataset directly to a TensorGraph and have it figure everything out?

**Comment by streetster17** [comment]
I'm new so I apoligize if I'm missing the direction of this thread but hopefully I can be of some assistance.

The HDF5 file format allows for hierarchical data storage and may help provide a clean way to tie multiple labels, weights, inouts and outputs to a single model. At first glance it seems pretty widely supported which may make it worthy of further invesitgation.

support.hdfgroup.org<http://support.hdfgroup.org>

Get Outlook for Android<https://aka.ms/ghei36>

**Comment by shntnu** [comment]
Our research focus is on assessing the impact of compounds and genetic perturbations on cells’ morphology, as imaged by microscopy, so I thought we'd chime in with some thoughts. 

We recently wrote a best practices [paper](https://www.ncbi.nlm.nih.gov/pubmed/28858338) together with nearly 20 other labs on the computational/statistical workflow for this kind of experiment. Most of that is focussed on "classical" approaches, i.e., not deep learning, but it would still be a good starting point if you want to learn about challenges unique to imaging data.

Our own lab (Imaging Platform/Carpenter lab at the Broad Institute), and several others, [have](https://github.com/broadinstitute/keras-rcnn) [been](https://github.com/broadinstitute/keras-resnet) [heavily](https://github.com/broadinstitute/deepometry) [invested](https://www.nature.com/articles/d41586-018-02174-z) in deep learning for image analysis, including image-based – or morphological – profiling. Our initial experiments [have](https://www.biorxiv.org/content/early/2017/12/02/227645) [been](https://www.biorxiv.org/content/early/2016/11/02/085118) promising, and likewise for some of [our](https://www.biorxiv.org/content/early/2017/07/10/161422) colleagues who have take a similar approach. We've also been writing a Python [library](https://github.com/jccaicedo/DeepProfiler) to support this research, and will soon post a Jupyter notebook to reproduce results from our [CVPR 2018](http://cvpr2018.thecvf.com/) paper (in press) in case you are interested in testing out the library. 

Based on our experience, we believe that for an application such as the Cell Chem Bio [paper](http://www.cell.com/cell-chemical-biology/fulltext/S2451-9456(18)30037-0), where the matrix of ground truth is typically sparse, it could be challenging to use a simple strategy of directly training a CNN to predict the response, given the images. Instead, we believe one should either use pre-trained networks, then fine tune, or directly use features that were extracted from images using classical image analysis techniques (as was done in the  [paper](http://www.cell.com/cell-chemical-biology/fulltext/S2451-9456(18)30037-0)). Or, one could train using weak labels – which is what our CVPR paper is about (we will post the link here in a few weeks) – and use the learned representation for the prediction tasks.

I hope that perspective helps when making decisions about your API. 

[@jccaicedo](https://github.com/jccaicedo) and @shntnu

**Comment by rbharath** [comment]
Added some first steps towards Image support here:

https://github.com/deepchem/deepchem/pull/1324

**Comment by rbharath** [comment]
I'm going to close this old issue since #1324 and other subsequent PRs have improved image handling support a good bit in DeepChem. Feel free to suggest re-opening if there's more discussion you'd like to see here.
